json-rpc/src/stream_rpc/connection/manager.rs

```diff
@@ -0,0 +1,159 @@
+        }
+    }
+
+    pub async fn client_connection(
+        self,
+        client_sender: StreamSender,
+        mut client_rcv: BoxConnectionStream,
+        connection_context: ConnectionContext,
+    ) {
+        let client_id = self.next_user_id();
+        let client = ClientConnection::new(
+            client_id,
+            client_sender.clone(),
+            connection_context,
+            self.config.clone(),
+        );
+        self.add_client(client.clone());
+
+        counters::CLIENT_CONNECTED
+            .with_label_values(&[
+                client.connection_context.transport.as_str(),
+                client.connection_context.sdk_info.language.as_str(),
+                &client.connection_context.sdk_info.version.to_string(),
+            ])
+            .inc();
+
+        debug!(
+            logging::ClientConnectionLog {
+                transport: client.connection_context.transport.as_str(),
+                remote_addr: client.connection_context.remote_addr.as_deref(),
+                client_id: Some(client.id),
+                user_agent: None,
+                forwarded: None,
+                rpc_method: None
+            },
+            "client connected"
+        );
+
+        // Reap client we can no longer send to it
+        let send_task = tokio::task::spawn(async move {
+            loop {
+                tokio::time::sleep(tokio::time::Duration::from_secs(5)).await;
+                if client_sender.is_closed() {
+                    break;
+                }
+            }
+        });
+
+        // TODO: reap idle connections without any subscriptions or which haven't accepted a message in a while?
+        let task_client = client.clone();
+        let task_db = self.get_db();
+        let recv_task = tokio::task::spawn(async move {
+            while let Some(result) = client_rcv.next().await {
+                match result {
+                    Ok(msg) => {
+                        if let Some(message) = msg {
+                            task_client.received_message(task_db.clone(), message).await;
+                        }
+                    }
+                    Err(e) => {
+                        debug!(
+                            logging::ClientConnectionLog {
+                                transport: task_client.connection_context.transport.as_str(),
+                                user_agent: None,
+                                remote_addr: task_client.connection_context.remote_addr.as_deref(),
+                                client_id: Some(task_client.id),
+                                forwarded: None,
+                                rpc_method: None
+                            },
+                            "client disconnect start {}", e
+                        );
+                        break;
+                    }
+                };
+            }
+        });
+
+        tokio::select! {
+            _ = send_task => (),
+            _ = recv_task => (),
```

Comment:
this is kind of a long function that basically just listens for either the sending connection or receiving connection to die and cleans up the connection then... perhaps we could state something like that within this code?